### PR TITLE
W7 - PhongLe - Unit Test - UI Test

### PR DESCRIPTION
**Pull request by at-phongle**
#### Related issues (Optional)
Task #92 

#### Description (Required)
- [x] Create UI Login
- [x] Write Validate
- [x] Write class unit test
- [ ] Write class test UI

### CODE REVIEW

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+import vn.asiantech.internship.models.Account;
+import vn.asiantech.internship.unittest.ValidatePassword;
+
+/**
+ * Created by phongle on 4/1/2561.
+ * Class UnitTest Password
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class PasswordUnitTest {
```
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    @Mock
+    private Account mAccount;
+    @Test
+    public void CheckDifferenceUsername(){
```
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        Mockito.when(mAccount.getUsername()).thenReturn("Phongle");
+        Mockito.when(mAccount.getPassword()).thenReturn("Phongle");
+        Assert.assertFalse(ValidatePassword.isDifferenceUsername(mAccount.getUsername(),mAccount.getPassword()));
+        Mockito.when(mAccount.getUsername()).thenReturn("Phonglet");
+        Mockito.when(mAccount.getPassword()).thenReturn("Phonglethanh");
+        Assert.assertTrue(ValidatePassword.isDifferenceUsername(mAccount.getUsername(),mAccount.getPassword()));
+    }
+    @Test
+    public void CheckLengthPassword(){
```
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        Mockito.when(mAccount.getPassword()).thenReturn("admin");
+        Assert.assertFalse(ValidatePassword.checkMinLength(mAccount.getPassword()));
+        Mockito.when(mAccount.getPassword()).thenReturn("adminPhong");
+        Assert.assertTrue(ValidatePassword.checkMinLength(mAccount.getPassword()));
+    }
+    @Test
+    public void CheckAtLeastSpecialCharOrNum(){
```
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        Mockito.when(mAccount.getPassword()).thenReturn("phongleth");
+        Assert.assertFalse(ValidatePassword.atLeastSpecialCharOrNumber(mAccount.getPassword()));
+        Mockito.when(mAccount.getPassword()).thenReturn("phongle1");
+        Assert.assertTrue(ValidatePassword.atLeastSpecialCharOrNumber(mAccount.getPassword()));
+        Mockito.when(mAccount.getPassword()).thenReturn("phongle@");
+        Assert.assertTrue(ValidatePassword.atLeastSpecialCharOrNumber(mAccount.getPassword()));
+    }
+    @Test
+    public void CheckAtThreeUppercase(){
```
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        Mockito.when(mAccount.getPassword()).thenReturn("phongleTH");
+        Assert.assertFalse(ValidatePassword.isAtLeastThreeUpperCase(mAccount.getPassword()));
+        Mockito.when(mAccount.getPassword()).thenReturn("phongTHA");
+        Assert.assertTrue(ValidatePassword.isAtLeastThreeUpperCase(mAccount.getPassword()));
+    }
+    @Test
+    public void CheckNonRepeat(){
```
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        Mockito.when(mAccount.getPassword()).thenReturn("lelelephong");
+        Assert.assertFalse(ValidatePassword.isNonRepeat(mAccount.getPassword()));
+        Mockito.when(mAccount.getPassword()).thenReturn("lephongle");
+        Assert.assertTrue(ValidatePassword.isNonRepeat(mAccount.getPassword()));
+    }
+    @Test
+    public void CheckNonSpace(){
```
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/UsernameUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+import vn.asiantech.internship.models.Account;
+import vn.asiantech.internship.unittest.ValidateUsername;
+
+/**
+ * Created by phongle on 4/1/2561.
+ * Class unit test username
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class UsernameUnitTest {
```
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/UsernameUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java (hunk 1 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java (hunk 2 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java (hunk 3 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java (hunk 4 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java (hunk 5 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java (hunk 6 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java (hunk 7 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/UsernameUnitTest.java (hunk 8 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/UsernameUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java (hunk 1 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java (hunk 2 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java (hunk 3 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java (hunk 4 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java (hunk 5 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java (hunk 6 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java (hunk 7 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/testAt-phongleDebug/java/UsernameUnitTest.java (hunk 8 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/UsernameUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

**Comment by at-phongle** [comment]
@at-trungnguyen  please review this . Thanks

File: app/src/testAt-phongleDebug/java/UsernameUnitTest.java (hunk 8 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/UsernameUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

File: app/src/at-phongle/java/vn/asiantech/internship/unittest/ValidateUsername.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+ * Class check validate username
+ */
+public final class ValidateUsername {
+    /**
+     * Check length username more than 5 and little than 24
+     *
+     * @param username username
+     * @return checkLengthUsername boolean state
+     */
+    public static boolean checkLengthUsername(String username) {
```
Comment (at-trungnguyen) [inline app/src/at-phongle/java/vn/asiantech/internship/unittest/ValidateUsername.java]: method start by `is`

File: app/src/at-phongle/java/vn/asiantech/internship/unittest/ValidateUsername.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+     * @param username username
+     * @return isNonWhiteSpace boolean state
+     */
+    public static boolean isNonWhiteSpace(String username) {
+        for (int i = 0; i < username.length(); i++) {
+            if (Character.isWhitespace(username.charAt(i))) {
+                return false;
+            }
+        }
+        return true;
```
Comment (at-trungnguyen) [inline app/src/at-phongle/java/vn/asiantech/internship/unittest/ValidateUsername.java]: `return !username.contains(" ");`

File: app/src/at-phongle/res/layout/activity_app.xml [hunk 11]
```diff
@@ -1,62 +1,81 @@
...
-    android:padding="10dp">
+    android:background="@drawable/bg_main">
 
-    <Button
-        android:id="@+id/btnLoginScreen"
+    <ScrollView
         android:layout_width="match_parent"
         android:layout_height="wrap_content"
-        android:text="@string/btn_login_screen" />
+        tools:ignore="UselessParent">
```
Comment (at-trungnguyen) [inline app/src/at-phongle/res/layout/activity_app.xml]: remove this

File: app/src/at-phongle/res/values/strings.xml [hunk 12]
```diff
@@ -60,4 +60,20 @@
     <string name="btn_ninepath">NinePath</string>
     <string name="btn_vector">Vector</string>
     <string name="btn_image_ninepath_vector">Image-Ninepath-Vector</string>
+    <string name="tv_username">Username :</string>
+    <string name="tv_password">Password :</string>
+    <string name="btn_login">Login</string>
+    <string name="check_length_username">Username phai lon hon 5 va ngan hon 24 ki tu !</string>
```
Comment (at-trungnguyen) [inline app/src/at-phongle/res/values/strings.xml]: use EL please

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+import vn.asiantech.internship.unittest.ValidatePassword;
+
+/**
+ * Created by phongle on 4/1/2561.
+ * Class UnitTest Password
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class PasswordUnitTest {
+    @Mock
+    private Account mAccount;
```
Comment (at-trungnguyen) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: break 1 line

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    @Test
```
Comment (at-trungnguyen) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: break 1 line

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+    public void checkDifferenceUsername(){
+        Mockito.when(mAccount.getUsername()).thenReturn("Phongle");
+        Mockito.when(mAccount.getPassword()).thenReturn("Phongle");
+        Assert.assertFalse(ValidatePassword.isDifferenceUsername(mAccount.getUsername(),mAccount.getPassword()));
+        Mockito.when(mAccount.getUsername()).thenReturn("Phonglet");
+        Mockito.when(mAccount.getPassword()).thenReturn("Phonglethanh");
+        Assert.assertTrue(ValidatePassword.isDifferenceUsername(mAccount.getUsername(),mAccount.getPassword()));
+    }
+    @Test
```
Comment (at-trungnguyen) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: break 1 line

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    public void checkLengthPassword(){
+        Mockito.when(mAccount.getPassword()).thenReturn("admin");
+        Assert.assertFalse(ValidatePassword.checkMinLength(mAccount.getPassword()));
+        Mockito.when(mAccount.getPassword()).thenReturn("adminPhong");
+        Assert.assertTrue(ValidatePassword.checkMinLength(mAccount.getPassword()));
+    }
```
Comment (at-trungnguyen) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: break 1 line

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+    @Test
+    public void checkAtLeastSpecialCharOrNum(){
+        Mockito.when(mAccount.getPassword()).thenReturn("phongleth");
+        Assert.assertFalse(ValidatePassword.atLeastSpecialCharOrNumber(mAccount.getPassword()));
+        Mockito.when(mAccount.getPassword()).thenReturn("phongle1");
+        Assert.assertTrue(ValidatePassword.atLeastSpecialCharOrNumber(mAccount.getPassword()));
+        Mockito.when(mAccount.getPassword()).thenReturn("phongle@");
+        Assert.assertTrue(ValidatePassword.atLeastSpecialCharOrNumber(mAccount.getPassword()));
+    }
```
Comment (at-trungnguyen) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: same above

File: app/src/testAt-phongleDebug/java/PasswordUnitTest.java [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+
+import vn.asiantech.internship.models.Account;
+import vn.asiantech.internship.unittest.ValidatePassword;
+
+/**
+ * Created by phongle on 4/1/2561.
+ * Class UnitTest Password
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class PasswordUnitTest {
```
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/PasswordUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

File: app/src/testAt-phongleDebug/java/UsernameUnitTest.java (hunk 8 above)
Comment (at-dev-support-01) [inline app/src/testAt-phongleDebug/java/UsernameUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

**Pull request merged**